app/controllers/links_controller.rb

```diff
@@ -9,10 +10,13 @@ class LinksController < ApplicationController
   before_action :authorize, only: %i[index new edit create destroy]
 
   # 2. set @link instance var, since a lot of action filters use it
-  before_action :set_link, only: %i[show edit update destroy export toggle_ability]
+  before_action only: %i[show edit update destroy export toggle_ability] do
+    set_link(params[:id])
+  end
+  before_action :require_link, only: %i[show edit update destroy export toggle_ability]
 
   # 3. protect link-specific buisness rules
-  before_action :prevent_public_expired, only: %i[show update]
+  before_action :prevent_public_expired, only: %i[show update], if: -> { link_is_expired_and_not_owned? }
```

Comment:
I wanted to move more of the logic itself into booleanish methods to make it easier to reuse; let me know if that's an antipattern for any reason